spec/unit/resource/rhsm_subscription_spec.rb

```diff
@@ -38,6 +43,46 @@
     expect { resource.action :remove }.not_to raise_error
   end
 
+  describe "#action_attach" do
+    before do
+      dummy = Mixlib::ShellOut.new
+      allow_any_instance_of(Chef::Mixin::ShellOut).to receive(:shell_out!).with("subscription-manager attach --pool=#{resource.pool_id}").and_return(dummy)
+      allow(dummy).to receive(:stdout).and_return("Successfully attached a subscription for: My Subscription")
+      allow(dummy).to receive(:exitstatus).and_return(0)
+      allow(dummy).to receive(:error?).and_return(false)
+      node.automatic_attrs[:platform_family] = "rhel"
+      node.automatic_attrs[:platform_version] = "7.3"
+      allow_any_instance_of(Chef::Provider::Package::Yum).to receive(:installed_version).with(0).and_return(Chef::Provider::Package::Yum::Version.new(nil, nil, nil))
+      allow_any_instance_of(Chef::Provider::Package::Yum).to receive(:available_version).with(0).and_return(Chef::Provider::Package::Yum::Version.new(nil, nil, nil))
+      allow_any_instance_of(Chef::Provider::Package::Yum::PythonHelper).to receive(:close_rpmdb)
```

Comment:
You're trying way too hard here.  You're testing that the details of the ResourceResolver correctly give back a Chef::Package::Provider::Yum when you're forcing the distro version, which is duplicating all the testing which goes on already to prove that happens.

I'd just create a double object and:

```
expect(provider).to receive(:build_resource).with(<stuff>).and_return(mydouble)
expect(mydouble).to receive(:run_action).with(:flush_cache)
```